### PR TITLE
Use CodeMirror editor instead of Monaco on non-desktop devices

### PR DESCRIPTION
**Pull request by Samywamy10**
# Problem

Recently my laptop has been in for repair and I've been using my iPad. One major issue I was finding with the config editor was an inability to select text.

https://github.com/blakeblackshear/frigate/assets/1314053/0618b3d1-601b-46b3-a2ad-83c432e05b66

## Solution
I looked into Monaco a bit to see if it was possible to fix this, but looks like a wider issue of mobile not being a priority: https://github.com/microsoft/monaco-editor/issues/246

Monaco obviously has a lot of pros still, so one solution suggested was to switch on non-desktop devices to [CodeMirror](https://codemirror.net/), who have invested in solving this problem with touch-based devices. CodeMirror isn't a small project - it has approx 1.9m weekly downloads (Monaco has 1.3m for reference).

So I've made a change to swap out the config editor with CodeMirror when on non-desktop devices. As you can see here it fixes my issue:

https://github.com/blakeblackshear/frigate/assets/1314053/aedda8ef-ed06-43a6-886a-0b03a1a18cd1

# Implementation
To enable this technically:
1. Abstracted the `CodeEditor` functionality into a new component with its own interface
2. Created two implementations of the `CodeEditor`: `MonacoEditor` and `CodeMirrorEditor`
3. Swap out the implementation depending on whether `isDesktop` is true or not

I've also removed some of the refs and replaced with state in the Monaco editor to follow [React best practices about using refs as an escape hatch](https://react.dev/learn/referencing-values-with-refs#best-practices-for-refs). I did a bit of testing around different situations the state could be in and this seems to work but let me know if I've missed a case and I'll have a look.

# Cons/discussion
At the moment the `CodeMirrorEditor` doesn't support YAML validation. From what I can gather we could implement this manually (CodeMirror supports validators, and there are many YAML validators out there), but I thought that'd expand the size of this PR and can be done in a follow up.

That leads me to my next (and probably your next) concern - do we want to be maintaining the code for two editors? CodeMirror is ultimately a very flexible and configurable editor, but with that comes a level of complexity in setting it up that Monaco makes a bit easier. I don't really have a good answer to this other than pointing to the popularity and strong community behind CodeMirror. But obviously there's a maintenance cost to this project in needing to support two editors, so will leave it to the discussion as to whether being able to edit the config file on touch devices is important enough or not.

### CODE REVIEW

**Comment by NickM-27** [comment]
Personally, I don't think maintaining two editors makes sense, and it increases the bundle size. Not sure about code mirror but monaco is pretty large.

In general, as we continue work on the UI it will trend towards most things being editable in the UI directly and not being done with yaml in the UI so this isn't a problem that will exist indefinitely. Personally I don't view doing config edits as much of a use case on mobile. iPad is an interesting crossroads there but via touch it just doesn't seem great. Open to see what other maintainers think

**Comment by Samywamy10** [comment]
> and it increases the bundle size

That's a good point. Just added lazy loading to the new CodeMirror implementation.

> In general, as we continue work on the UI it will trend towards most things being editable in the UI 

Yup, I've liked seeing the progress on the UI here. I'd be more than happy to remove this implementation when we get to this point!

**Comment by hawkeye217** [comment]
I'm with @NickM-27, I don't think maintaining two editors is worthwhile at this point - Monaco has been a part of Frigate for a while and I don't recall any users making comments about the lack of functionality on mobile devices.

We definitely should aim to keep the bundle size down as much as possible.

**Comment by Samywamy10** [comment]
Up to you. Bundle size won't be affected given lazy loading fwiw.

Shall I make an issue with the touch selection issues I'm having?

File: web/src/components/code_editor/CodeEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+/** Used to swap between the Monaco editor and the CodeMirror editor depending on device type.
+ * The Monaco editor doesn't work well on mobile devices, see Github issue: https://github.com/microsoft/monaco-editor/issues/246
+ *
+ * A common solution is to switch to CodeMirror on mobile devices as they have invested significantly in touch support.
+ *
+ * It would be great to revisit this in in the future if/when Monaco improves mobile support.
+ */
+export const CodeEditor: (props: CodeEditorProps) => React.ReactNode = isDesktop
+  ? MonacoEditor
+  : lazy(() => import("./CodeMirrorEditor"));
```
Comment (Samywamy10) [inline web/src/components/code_editor/CodeEditor.tsx:23]: Can add lazy loading to the Monaco Editor aswell

**Comment by T3chArmy** [comment]
I personally use my mobile devices quite a bit to edit config, and I have to agree with Samy, it's been a MASSIVE pain point for me in the past. I would say that I have been using the config editor significantly less with 0.14 Betas though.

**Pull request closed without merging**